### ISSUE TITLE
False positives for CSS3 namespaces in selector-combinator-space-*

### ISSUE THREAD

**Issue by bedney** [stylelint/stylelint#2711]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A bug where this rule, which is perfectly valid CSS3 with a namespace prefix, is being marked as having an issue:

`sherpa|tile#StyleSummary_Tile > .body {...}`

This reports the following error:

`76:31  ✖  Expected single space after ">"    selector-combinator-space-after`

There is, indeed, a space after the `>` (I checked to make sure it wasn't a Tab)

> Which rule, if any, is this issue related to?

`selector-combinator-space-after`

> What CSS is needed to reproduce this issue?

e.g.

```css
sherpa|tile#StyleSummary_Tile > .body {...}
```

> What stylelint configuration is needed to reproduce this issue?

e.g.

```json
{
    "extends": "stylelint-config-standard",
    "rules": {
        "at-rule-empty-line-before": null,

        "comment-empty-line-before": null,

        "block-no-empty": null,

        "block-opening-brace-newline-after": "always",

        "declaration-empty-line-before": null,

        "indentation": 4,

        "font-family-name-quotes": "always-unless-keyword",

        "function-url-quotes": "always",

        "rule-nested-empty-line-before": null,

        "selector-attribute-quotes": "always",

        "selector-type-no-unknown": [ true, {
            "ignoreNamespaces": [ "tibet", "tsh", "sherpa", "xctrls" ]
        }],

        "selector-type-case": null,

        "string-quotes": "double",

        "value-keyword-case": "lower"
    },
    "ignoreFiles": [
        "node_modules/**",
        "**/node_modules/**",
        "deps/**",
        "lib/src/**",
        "**/sherpa/cm_theme/*",
        "**/build/**",
        "**/*.__*__.*.js",
        "**/boot/styles/**"
    ]
}
```

> Which version of stylelint are you using?

`7.12.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

"CLI with `stylelint mycss.css`"

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No, regular, W3C-approved CSS3.

> What did you expect to happen?

"No warnings to be flagged."

> What actually happened (e.g. what warnings or errors you are getting)?

"The following warnings were flagged:"

```shell
76:31  ✖  Expected single space after ">"    selector-combinator-space-after
```

**Comment by bedney** [comment]
One last thing: You can see that the bug with this rule goes away if you remove either the `sherpa|tile` or the `#StyleSummary_Tile` portion of the selector.

To avoid fixing these piecemeal, I would make a suggestion to add at selector that has a CSS3 namespace prefix to least one test case per style lint rule. I have a feeling that this issue is due to this particular rule not anticipating the `|` of CSS3 namespaces. And I also have a suspicion that this assumption was made in other rules as well.

**Comment by jeddy3** [comment]
@bedney Thank you for the report and for using the issue template.

I can replicate this for both the `selector-combinator-space-*` rules. Fortunately, the 2 other `*-combinator(s)` rules are fine. 

The bug is likely in [`selectorCombinatorSpaceChecker.js`](https://github.com/stylelint/stylelint/blob/master/lib/rules/selectorCombinatorSpaceChecker.js).

Labeling as `help wanted` and `bug`. @bedney Feel free to [contribute a fix](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#fixing-a-bug-in-an-existing-rule).

**Comment by bedney** [comment]
Yea, it's almost like these rules aren't recognizing that `|` can be a valid character in a CSS selector. I'm not sure I'm going to be of much help... I wish I was more of a parser guy... sigh.

Thanks for taking a look.

**Comment by jeddy3** [comment]
This is fixed. The issue uncovered a bug in `postcss-selector-parser` [that'll be fixed in `3.0`](https://github.com/postcss/postcss-selector-parser/issues/111). However, a slight change in stylelint code means we no longer make use of that feature in this instance. So, we don't need to wait for that release.

> I'm not sure I'm going to be of much help... I wish I was more of a parser guy... sigh.

Me too... :) stylelint itself consumes various parsers and most of working with stylelint is about using the (pleasant) APIs that these parsers provide: [PostCSS API](http://api.postcss.org/) & [postcss-selector-parser API](https://github.com/postcss/postcss-selector-parser/blob/master/API.md). So, it's pretty high-level, and as such it's not too daunting and it's pretty easy to get stuck in. I mention this because we're always in need of new contributors to keep the project afloat, and any help is appreciated :)

**Comment by bedney** [comment]
Wow! Thanks for the bug fix! And thank you for reporting it so that it could be fixed upstream.

I could help by adding to your tests to include CSS3 namespace constructs where appropriate (both element-level and attribute-level). Does that sound helpful?

Thanks again! This is the most responsive open source project I ever seen :-).

**Comment by ntwb** [comment]
> I could help by adding to your tests to include CSS3 namespace constructs where appropriate (both element-level and attribute-level). Does that sound helpful?

It would be better than helpful, it would be awesome @bedney, the more coverage the better 😄
